@here/harp-datasource-protocol/lib/ExprEvaluator.ts

```diff
@@ -50,19 +51,66 @@ const operatorDescriptors = new Map<string, OperatorDescriptor>();
  * @hidden
  */
 export class ExprEvaluatorContext {
+    private _partialEvaluation = false;
+
     constructor(
         readonly evaluator: ExprEvaluator,
         readonly env: Env,
         readonly scope: ExprScope,
         readonly cache?: Map<Expr, Value>
     ) {}
 
+    /**
+     * `true` if the this context is used to partially evaluate expressions.
+     */
+    get partialEvaluation() {
+        return this._partialEvaluation;
+    }
+
+    /**
+     * Evaluate the given expression.
+     *
+     * @param expr The [[Expr]] to evaluate.
+     */
     evaluate(expr: Expr | undefined) {
         if (expr !== undefined) {
             return expr.accept(this.evaluator, this);
         }
         throw new Error("Failed to evaluate expression");
     }
+
+    /**
+     * Partially evaluate the given expression.
+     *
+     * @param expr The [[Expr]] to evaluate.
+     */
+    partiallyEvaluate(expr: Expr | undefined): Expr {
+        if (expr === undefined) {
+            throw new Error("Failed to evaluate expression");
```

Comment:
Why accept `undefined` if it should not be supported?

Reply:
like for the above `evaluate`, possible `undefined` values may come for the recursive evaluation of expressions.